servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java

```diff
@@ -266,7 +266,8 @@ public String toString() {
             this.splittingFlushStrategy = new SplittingFlushStrategy(connection.defaultFlushStrategy(),
                     itemWritten -> {
                         if (itemWritten instanceof HttpResponseMetaData) {
-                            return Start;
+                            final HttpResponseMetaData metadata = (HttpResponseMetaData) itemWritten;
+                            return emptyMessageBody(metadata) && protocol().major() > 1 ? End : Start;
```

Comment:
nit: consider flipping the order of evaluation to short circuit on `protocol().major() > 1` which should be the less expensive to evaluate.

Reply:
https://github.com/apple/servicetalk/pull/1679#discussion_r672638667

> @Scottmitch I understand that a different format of the flattened HTTP message needs additional coordination for places that expect trailers at the end. However, the perf bump of reducing the number of h2 frames is significant enough (see description of #1518) to justify these adjustments.

It is not ideal to have different semantics on the flattened stream depending upon protocol. The `{headers, payload*, trailers}` sequencing was devised bcz we couldn't inspect the stream contents and didn't have the meta data retained in `PayloadInfo`. Another option is to use the `LastHttpChunk` interface style approach (we use this in Netty) to delineate message boundaries. This may incur object allocation overhead to wrap, but would may make the internal ST<->http transport API more explicit (and agnostic to protocol version).

> Also, I hope that this SplittingFlushStrategy logic will go away after we rework the server-side internals.

me too 🙏

Reply:
I would propose to defer alternative approaches because:
1. They will require a lot more refactoring at different layer, it may be easy to miss something.
2. Extra allocation and wrapping/unwrapping is not desirable.
3. If `SplittingFlushStrategy` will go away, we won't need any marker for the last chunk because it can be managed by publisher completion.

Reply:
yip fine to defer, lets just make sure this planned work is tracked somewhere (with reference back to this discussion to restore context if necessary).